### PR TITLE
nixos/llama-cpp: init

### PR DESCRIPTION
**Pull request by newAM**
## Description of changes

Add a module for the `llama-cpp` WebUI.

For testing you can use this to build `Mistral-7B-Instruct-v0.2` with `q4_0` quantization from source:

> [!WARNING]  
> The download is around ~30GB.  Evaluating this uses a lot of memory I do not recommend evaluating it unless you have at least 64GB system memory.

```nix
{
  stdenvNoCC,
  fetchgit,
  llama-cpp,
  python3,
}:
stdenvNoCC.mkDerivation rec {
  pname = "mistral-instruct-7b";
  version = "0.2";

  src = fetchgit {
    url = "https://huggingface.co/mistralai/Mistral-7B-Instruct-v${version}";
    rev = "b70aa86578567ba3301b21c8a27bea4e8f6d6d61";
    hash = "sha256-IAe/tHFB7yqFRF5aRojkNCD8TbKj8XQMt6eEyPmr4HU=";
    fetchLFS = true;
  };

  nativeBuildInputs = [
    llama-cpp
    python3
    python3.pkgs.numpy
    python3.pkgs.sentencepiece
    python3.pkgs.transformers
  ];

  buildPhase = ''
    mkdir $out
    python3 ${llama-cpp.src}/convert.py $src --outfile /build/ggml-model-f16.gguf --outtype f16
    llama-cpp-quantize /build/ggml-model-f16.gguf $out/ggml-model-q4_0.gguf q4_0
  '';

  dontInstall = true;
  dontFixup = true;
}
```

CC `llama-cpp` package maintainers: @dit7ya @elohmeier

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/llama-cpp.nix [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      package = lib.mkPackageOption pkgs "llama-cpp" { };
+
+      model = lib.mkOption {
+        type = lib.types.path;
+        example = "/models/mistral-instruct-7b/ggml-model-q4_0.gguf";
+        description = lib.mdDoc "Model path.";
+      };
+
+      extraFlags = lib.mkOption {
+        type = lib.types.str;
```
Comment (jian-lin) [inline nixos/modules/services/misc/llama-cpp.nix:21]: More composable
```suggestion
        type = lib.types.listOf lib.types.str;
```

File: nixos/modules/services/misc/llama-cpp.nix [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    systemd.services.llama-cpp = {
+      description = "LLaMA C++ server";
+      after = ["network-online.target"];
+      wantedBy = ["multi-user.target"];
+
+      serviceConfig = {
+        Type = "idle";
+        KillSignal = "SIGINT";
+        ExecStart = "${cfg.package}/bin/llama-cpp-server --log-disable --host ${cfg.host} --port ${builtins.toString cfg.port} -m ${cfg.model} ${cfg.extraFlags}";
```
Comment (jian-lin) [inline nixos/modules/services/misc/llama-cpp.nix:61]:
```suggestion
        ExecStart = "${cfg.package}/bin/llama-cpp-server --log-disable --host ${cfg.host} --port ${builtins.toString cfg.port} -m ${cfg.model} ${utils.escapeSystemdExecArgs cfg.extraFlags}";
```

File: nixos/modules/services/misc/llama-cpp.nix [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    systemd.services.llama-cpp = {
+      description = "LLaMA C++ server";
+      after = ["network-online.target"];
```
Comment (jian-lin) [inline nixos/modules/services/misc/llama-cpp.nix:55]: This is usually enough for servers.  See `man systemd.special`.

```suggestion
      after = ["network.target"];
```

**Comment by newAM** [comment]
Thanks for the review @jian-lin!  I applied your feedback.

File: nixos/modules/services/misc/llama-cpp.nix [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [
+          "@system-service"
+          "~@debug"
+          "~@mount"
+          "~@privileged"
+          "~@resources"
+          "~@cpu-emulation"
+          "~@obsolete"
```
Comment (jian-lin) [inline nixos/modules/services/misc/llama-cpp.nix:99]: Sorry, I should have mentioned that many `~@foo` are redundant after adding `@system-service` to the front.  You can check that by running `systemd-analyze security llama-cpp`.

Leaving some redundant ones here may act as a not-so-good example for future contributors writing hardening options.

File: nixos/modules/services/misc/llama-cpp.nix [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    systemd.services.llama-cpp = {
+      description = "LLaMA C++ server";
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+
+      serviceConfig = {
+        Type = "idle";
+        KillSignal = "SIGINT";
+        ExecStart = "${cfg.package}/bin/llama-cpp-server --log-disable --host ${cfg.host} --port ${builtins.toString cfg.port} -m ${cfg.model} ${utils.escapeSystemdExecArgs cfg.extraFlags}";
```
Comment (jian-lin) [inline nixos/modules/services/misc/llama-cpp.nix:59]: I am not familiar with this specific application, however, disabling logging unconditionally seems a bit uncommon.  If users may can to enable logging, it is better to make an option for that.

File: nixos/modules/services/misc/llama-cpp.nix (hunk 4 above)
Comment (newAM) [reply nixos/modules/services/misc/llama-cpp.nix]: `@system-service` did not include `~@privileged` and `~@resources` so I left those in, but removed the others.  Thanks for letting me know about this!

File: nixos/modules/services/misc/llama-cpp.nix (hunk 5 above)
Comment (newAM) [reply nixos/modules/services/misc/llama-cpp.nix:59]: `--log-disable` is a bit of a misnomer here.  It disables logging to a file, and instead logs to stdout which gets captured by the journal.

**Comment by newAM** [comment]
Thanks for the reviews @jian-lin! :)

**Pull request merged**